### PR TITLE
feat: Client-side profile recommendation engine for Capacitor/direct mode

### LINKED ISSUE #398: feat: Client-side profile recommendation engine + structural auto-tagging
## Problem

1. **Bug**: "Find Similar Profiles" returns empty in Capacitor/direct mode — the recommendation engine lives in Python backend only, and `DirectModeInterceptor` stubs both `/api/profiles/find-similar` and `/api/profiles/recommend` with `{ recommendations: [] }` (lines 1931-1939).

2. **Missing feature**: Most profiles have no tags — only profiles created through the UI wizard have tags (stored as unstructured text in `user_preferences`). Imported profiles, machine-default profiles, and profiles created outside Metic have zero tags.

3. **Poor discoverability**: The profile catalogue tag filtering is nearly useless because so few profiles have extractable tags.

### PR 1: Client-Side Recommendation Engine (`feat/client-side-recommendations`)

Port the deterministic recommendation engine from Python to TypeScript:

- **`profileAnalysis.ts`** — Unified analysis core. Ports `_extract_fingerprint()` and `_extract_name_tags()` from `profile_recommendation_service.py`. Replaces 3 divergent analyzers (Python recommender, `staticProfileDescription.ts`, `ProfileBreakdown.tsx`).
- **`profileRecommendation.ts`** — Client-side scoring engine. Same algorithm: structure 35%, tags 25%, weight 15%, pressure 15%, temperature 10%. Pure functions, no caching.
- **`DirectModeInterceptor.ts`** — Replace empty stubs with calls to client-side engine.
- **Parity tests** — Golden fixtures matching Python `test_recommendations.py`.

### PR 2: Profile Auto-Tagging (`feat/profile-tags`)

- Derive structural tags **on-demand** from profile fingerprints (Pre-infusion, Bloom, Italian, Modern, Lever, Turbo, Long, Short)
- Enhance catalogue filtering to use derived + overlay tags
- Fold derived tags into recommendation scoring (Jaccard similarity)

Key design decisions:
- Structural tags computed on-demand, never persisted (survives profile edits)
- Only user/AI overlay tags are persisted
- One unified analysis core avoids divergent analyzers

## Related Issues
- AI tag enrichment deferred to v2.5.0 (see separate issue)
- Feature parity audit of DirectModeInterceptor (see separate issue)

### PR DESCRIPTION
**Pull request by hessius**
## Summary

Port the Python recommendation engine to TypeScript so "Find Similar Profiles" and tag-based recommendations work in the native iOS app (Capacitor/direct mode) without requiring the Python backend.

## Problem

In Capacitor/direct mode, `DirectModeInterceptor` stubbed both `/api/profiles/find-similar` and `/api/profiles/recommend` with empty `{ recommendations: [] }` responses. The feature was flagged as enabled in `featureFlags.ts` but silently returned no results.

### New files (1,092 lines added)

- **`profileAnalysis.ts`** — Unified profile structural analysis module. Faithful TypeScript port of Python `_extract_fingerprint()` and `_extract_name_tags()` from `profile_recommendation_service.py`. Extracts control mode, technique tags, peak pressure, temperature, weight, and structural features from profile stages.

- **`profileRecommendation.ts`** — Client-side scoring engine. Same algorithm as Python backend: structure (35%), tags (25%), weight (15%), pressure (15%), temperature (10%). Pure functions, no caching. Exports `findSimilarProfiles()` and `getRecommendations()`.

- **`profileAnalysis.test.ts`** + **`profileRecommendation.test.ts`** — 65 parity tests using the same 5 fixture profiles as Python `test_recommendations.py`. Validates fingerprint extraction, name tag extraction, scoring, ranking, and edge cases.

### Modified files

- **`DirectModeInterceptor.ts`** — Replaced empty stubs with client-side engine calls. On recommendation requests, fetches full profile list from machine via `/api/v1/profile/list`, runs scoring, returns results.

## Testing

- ✅ 786 frontend tests pass (65 new + 721 existing)
- ✅ Build succeeds
- ✅ App built and deployed to iPhone 14

## Related

- Closes #398
- See #399 (horizontal feature parity audit)
- See #400 (AI tag enrichment, deferred to v2.5.0)

### CODE REVIEW

**Review by hessius — COMMENTED** [review]
Addressed all 8 review findings in commit 632def77:

1. **App.tsx fallback** — Added fallback to fetch `/api/v1/profile/list` and resolve profile ID by name when cache miss occurs. Now handles both cache-hit and cache-miss paths.

2. **ProfileRecommendations conditional interactivity** — Card `role`, `tabIndex`, keyboard handlers, and `aria-label` are now only applied when `onUseProfile` callback is provided. No interactive ARIA attributes when non-clickable.

3. **FindSimilarOverlay conditional interactivity** — Same treatment applied. Interactive attributes conditional on `onSelectProfile` being provided.

4. **useSoundEffects docs** — Updated header docs: `back`/`close` → `click()` (was documented as `notify` but intentionally changed per user feedback to match forward navigation sound).

5. **buildUserFingerprint mappings** — Added `pressure-controlled`, `flow-controlled`, `mixed-controlled` → internal technique tag mappings so PRESET_TAGS selections properly influence scoring.

6. **Tag vocabulary normalization** — Internal technique tags (`pressure-profile`, `flow-profile`, etc.) are now normalized to user-facing labels before Jaccard comparison. Also added derived control-mode and temperature-range tags to candidate tag set for broader matching.

7. **Limit validation** — Both find-similar and recommend endpoints now validate/clamp `limit` to `[1, 100]` with `Number.isFinite()` guard.

8. **Profile list caching** — Added 30-second TTL in-memory cache via `_getCachedProfileList()`. Both recommendation endpoints now use this cached helper instead of fetching on every request.

File: apps/web/src/components/ProfileRecommendations.tsx [hunk 1]
```diff
@@ -161,53 +213,42 @@ export function ProfileRecommendations({
...
-                          {Math.round(rec.score)}
-                        </span>
+                  <Card
+                    className="p-2 sm:p-3 transition-colors cursor-pointer hover:bg-secondary/40"
+                    onClick={() => onUseProfile?.(rec.profile_name)}
+                    role="button"
+                    tabIndex={0}
+                    onKeyDown={(e) => { if (e.key === 'Enter' || e.key === ' ') { e.preventDefault(); onUseProfile?.(rec.profile_name) } }}
+                    aria-label={t('a11y.useProfile', { name: rec.profile_name })}
+                  >
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/components/ProfileRecommendations.tsx:225]: Already handled in the current code. The Card's interactive props (`role`, `tabIndex`, `aria-label`, `onClick`, `onKeyDown`, and `cursor-pointer`) are conditionally spread via `{...(onUseProfile ? { ...interactiveProps } : {})}` — they are only applied when `onUseProfile` is provided.

File: apps/web/src/components/FindSimilarOverlay.tsx [hunk 2]
```diff
@@ -145,45 +187,28 @@ export function FindSimilarOverlay({
                   >
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/components/FindSimilarOverlay.tsx:189]: Already addressed — same conditional spread pattern as ProfileRecommendations. The Card only gets interactive attributes when `onSelectProfile` is provided.

File: apps/web/src/hooks/useSoundEffects.ts [hunk 3]
```diff
@@ -143,7 +143,7 @@ export function useGlobalSoundDelegation() {
       switch (soundAttr) {
         case 'back':
         case 'close':
-          tiks.notify()
+          tiks.click()
           break
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/hooks/useSoundEffects.ts:147]: Fixed — updated the file-level JSDoc to say `"back", "close" → click` to match the actual `tiks.click()` implementation. The function-level JSDoc was already correct.

File: apps/web/src/lib/profileRecommendation.ts [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    pulse: 'pulse',
+    lever: 'lever',
+    turbo: 'turbo',
+    ramp: 'ramp',
+    decline: 'decline',
+    taper: 'taper',
+    flat: 'flat',
+    pressure: 'pressure-profile',
+    flow: 'flow-profile',
+  }
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/lib/profileRecommendation.ts:251]: Fixed. The tag label mappings were already present in `tagToTechnique`, but the `controlMode` assignment was missing the `mixed` case. Added `else if (techniqueTags.has('mixed-profile')) controlMode = 'mixed'` so the `Mixed-controlled` tag now correctly influences control-mode scoring.

File: apps/web/src/lib/profileRecommendation.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    const tagPts = tagSim * 25
+    score += tagPts
+
+    const overlap: string[] = []
+    for (const t of userLower) {
+      if (allCandTags.has(t)) overlap.push(t)
+    }
+    if (overlap.length > 0) {
+      // Filter out already-reported technique tags
+      const userTechTags = userFingerprint?.techniqueTags ?? new Set<string>()
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/lib/profileRecommendation.ts]: Already addressed. The scoring function normalizes internal technique tags to user-facing labels via `techniqueToLabel` (lines 149–156), and adds derived `${controlMode}-controlled` and temperature-range tags to `allCandTags` (lines 157–168).

File: apps/web/src/services/interceptor/DirectModeInterceptor.ts [hunk 6]
```diff
@@ -1926,16 +1928,59 @@ export function installDirectModeInterceptor(): void {
...
     if (url.match(/\/api\/profiles\/find-similar/) && method === 'POST') {
-      return Promise.resolve(jsonResponse({ recommendations: [] }))
+      return (async () => {
+        try {
+          let profileName = ''
+          let limit = 10
+          const body = init?.body
+          if (body instanceof FormData) {
+            profileName = body.get('profile_name')?.toString() ?? ''
+            limit = parseInt(body.get('limit')?.toString() ?? '10', 10)
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/services/interceptor/DirectModeInterceptor.ts]: Already addressed. Both endpoints validate/clamp `limit` with `Number.isFinite(parsed)` guard and `Math.min(100, Math.max(1, parsed))`, defaulting to 10 or 5 when parsing fails.

File: apps/web/src/services/interceptor/DirectModeInterceptor.ts [hunk 7]
```diff
@@ -1926,16 +1928,59 @@ export function installDirectModeInterceptor(): void {
...
+          }
+
+          const res = await _fetch('/api/v1/profile/list')
+          if (!res.ok) return jsonResponse({ recommendations: [] })
+          const profiles: AnalyzableProfile[] = await res.json()
+          if (!profiles?.length) return jsonResponse({ recommendations: [] })
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/services/interceptor/DirectModeInterceptor.ts]: Already addressed — a short-lived in-memory cache (`_getCachedProfileList()` with 30-second TTL) was added so both endpoints share cached data and avoid repeated machine calls on rapid tag changes.

File: apps/web/src/App.tsx [hunk 8]
```diff
@@ -806,18 +806,62 @@ function App() {
...
+          if (cacheData?.profile?.id) {
+            profileId = cacheData.profile.id
+            displayImage = cacheData.profile.display?.image
+          }
+        }
+
+        // Fetch profile JSON for the breakdown view
+        const jsonRes = await fetch(`/api/machine/profile/${encodeURIComponent(profileId)}/json`)
+        const jsonData = jsonRes.ok ? await jsonRes.json() : {}
+        const profileJson = jsonData?.profile ?? null
```
[the remark this answers is not in the document]
Comment (hessius) [reply apps/web/src/App.tsx]: Already addressed. The fallback to `/api/v1/profile/list` is present (lines 830–840). When the cache doesn't resolve an ID (`profileId === profileName`), it fetches the full profile list from the machine and resolves by name.
